Fix packet flood attack

FazeNextuss opened the pull request:
## Introduction
- This fix aims to block malicious attacks (flood of packets) by adding a limit. -

## Fix
- Added a packet limit sent by a player within a given time. (More than 5 packets in 2 seconds). If the player exceeds this limit, he will be kicked for "Too fast packet". 

## Test
- This has been tested, and now blocks these attacks.

FazeNextuss committed:
Fix packet flood attack

This fix aims to block malicious attacks (flood of packets) by adding a limit.

FazeNextuss committed:
Fix error

FazeNextuss committed:
fix error 2

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

DaisukeDaisuke commented on src/network/mcpe/NetworkSession.php:
```diff
@@ -179,7 +178,7 @@ class NetworkSession{
 	/** @var Packet[] */
 	private array $sendBuffer = [];
-
+	private array $packetsReceived = [];
```
```suggestion
	/** @var array<string, array{count: int, time: int}> */
	private array $packetsReceived = [];
```

DaisukeDaisuke commented on src/network/mcpe/NetworkSession.php:
```diff
@@ -419,10 +418,31 @@ public function handleEncoded(string $payload) : void{
+			];
+		}
+
+		$playerPackets = &$this->packetsReceived[$playerName];
+		$playerPackets['count']++;
```
I think passing by reference causes confusion
(It can be split into two properties)

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

DaisukeDaisuke commented on src/network/mcpe/NetworkSession.php:
```diff
@@ -419,10 +418,31 @@ public function handleEncoded(string $payload) : void{
 			throw new PacketHandlingException("Unexpected non-serverbound packet");
 		}
 
+		if (strlen($buffer) > 16500 && $packet->getName() === "InventoryTransactionPacket") {
```
```php
use pocketmine\network\mcpe\protocol\DisconnectPacket;
use pocketmine\network\mcpe\protocol\EmotePacket;
use pocketmine\network\mcpe\protocol\InventoryTransactionPacket; // new import
use pocketmine\network\mcpe\protocol\MobArmorEquipmentPacket;
```
```suggestion
		if (strlen($buffer) > 16500 && $packet instanceof InventoryTransactionPacket) {
```

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss commented:
Thanks for your help.
Right now correct my mistakes.

NhanAZ commented on src/network/mcpe/NetworkSession.php:
```diff
@@ -419,10 +419,31 @@ public function handleEncoded(string $payload) : void{
 			throw new PacketHandlingException("Unexpected non-serverbound packet");
 		}
 
+		if (strlen($buffer) > 16500 && $packet->getName() === "InventoryTransactionPacket") {
+			$this->logger->debug("Huge InventoryTransactionPacket: " . base64_encode($buffer));
+			throw new PacketHandlingException("InventoryTransactionPacket too big");
+		}
+		if (strlen($buffer) > 31968) {
+			if (time() - $playerPackets['time'] >= 2 && $playerPackets['count'] > 5) {
```
`16500`, `31968`
magic number

DaisukeDaisuke commented on src/network/mcpe/NetworkSession.php:
```diff
@@ -419,10 +419,31 @@ public function handleEncoded(string $payload) : void{
+		}
+		if (strlen($buffer) > 31968) {
+			if (time() - $playerPackets['time'] >= 2 && $playerPackets['count'] > 5) {
+				$this->getPlayer()?->kick("Too fast packet");
```
Here there is no interruption
```suggestion
				throw new PacketHandlingException("Too fast packet");
```

FazeNextuss committed:
Replace "kick" with "PacketHandlingException"

Replace "kick" with "PacketHandlingException" for there to be an interruption.

Co-authored-by: DaisukeDaisuke <17798680+DaisukeDaisuke@users.noreply.github.com>

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

DaisukeDaisuke commented on src/network/mcpe/NetworkSession.php:
```diff
@@ -419,10 +419,31 @@ public function handleEncoded(string $payload) : void{
 	public function handleDataPacket(Packet $packet, string $buffer) : void{
+
+		$playerName = $this->getPlayer()?->getName();
+		if (!isset($this->packetsReceived[$playerName])) {
+			$this->packetsReceived[$playerName] = [
+				'count' => 0,
+				'time' => time(),
+			];
+		}
```
After initializing the `packetsReceived` property, the `time` appears unchanged
```
array(2) {
  ["count"]=>
  int(6557)
  ["time"]=>
  int(1674995763)
}
```

FazeNextuss committed:
add InventoryTransactionPacket

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

FazeNextuss committed:
Update NetworkSession.php

dktapps commented:
There is already a limit.

Viska2007 commented:
Hello, I just saw that you just closed my pull request. But your limit doesn't work. The person who has the code that allows down the servers with what I did is now impossible. You what you have done is a basic "limit" but what it does is that it sends a lot of requests in a given time. What I have done is that in 2 seconds, it cannot send 5 packets of more than approximately 30000 bits. After having tested this flaw by having put my fix, it no longer works.

## Here is the packet which is spammy and quite big which freezes the server (encrypted):
Unhandled PlayerAuthInputPacket: kAEAAMB/AADAfwAAtUP0PetCAICRQwAAAAAAAAAAAADAf4CAgIiAAQECAgAAAAAAAAAAAAAAAACg/

## By putting my fix:
![devfix9999](https://user-images.githubusercontent.com/106326833/215330745-76f1e263-3b06-4951-91dc-31579a54cb47.png)
